marshal_3_smallint_test.go

```diff
@@ -0,0 +1,109 @@
+package gocql
+
+import (
+	"testing"
+
+	"github.com/gocql/gocql/marshal/tests/funcs"
+	"github.com/gocql/gocql/marshal/tests/mod"
+	"github.com/gocql/gocql/marshal/tests/serialization"
+)
+
+func TestMarshalSmallint(t *testing.T) {
+	marshal := func(i interface{}) ([]byte, error) { return Marshal(NativeType{proto: 4, typ: TypeSmallInt}, i) }
+	unmarshal := func(bytes []byte, i interface{}) error {
+		return Unmarshal(NativeType{proto: 4, typ: TypeSmallInt}, bytes, i)
+	}
```

Comment:
Let's discuss possibility of handling cases when it suppose to fail in the following manner:
```
	unmarshal := func(bytes []byte, i interface{}) (err error) {
		defer func() {
			if r := recover(); r != nil {
				err = utils.PanicErr{Err: r.(error), Stack: debug.Stack()}
			}
			rVal := reflect.ValueOf(i)
			if rVal.Kind() != reflect.Ptr {
				err = utils.CatchExpectedError("can not unmarshal into non-pointer", err)
			} else if rVal.IsNil() {
				switch reflect.TypeOf(i).Elem().Kind() {
				case reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64, reflect.Int, reflect.String,
					reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64, reflect.Uint:
					err = utils.CatchExpectedError("invalid memory address or nil pointer dereference", err)
				}
			}
		}()
		return Unmarshal(NativeType{proto: 4, typ: TypeSmallInt}, bytes, i)
	}
```

where `CatchExpectedError` is defined as such:
```
type UnExpectedSuccess struct {
	ActualErr   error
	ExpectedErr string
}

func (e *UnExpectedSuccess) Error() string {
	return fmt.Sprintf("unexpected success, expected error %q but got %v", e.ExpectedErr, e.ActualErr)
}

type ExpectedFailure struct {
	ActualErr   error
	ExpectedErr string
}

func (e *ExpectedFailure) Error() string {
	return fmt.Sprintf("failed as expected with error %v", e.ActualErr)
}

type UnExpectedFailure struct {
	ActualErr   error
	ExpectedErr string
}

func (e *UnExpectedFailure) Error() string {
	return fmt.Sprintf("failed as expected with error %v", e.ActualErr)
}

func CatchExpectedError(expectedErr string, actualErr error) error {
	if actualErr == nil {
		return &UnExpectedSuccess{ActualErr: actualErr, ExpectedErr: expectedErr}
	}
	if !strings.Contains(actualErr.Error(), expectedErr) {
		return &UnExpectedFailure{ExpectedErr: expectedErr, ActualErr: actualErr}
	}
	return &ExpectedFailure{ExpectedErr: expectedErr, ActualErr: actualErr}
}

func IsExpectedFailure(err error) bool {
	_, ok := err.(*ExpectedFailure)
	return ok
}
```

and error is being handled like this:
```
		err := func() (err error) {
			defer utils.CatchPanic(&err)
			return f(bytes.Clone(s.Data), got)
		}()
		if err != nil {
			if utils.IsExpectedFailure(err) {
				return
			}
			t.Fatalf("unmarshal unexpectedly failed with error: %s", err.Error())
		}

```

Reply:
Let's finish with the `serialization.Set`, and then we will continue this discussion. We have A lot of changes.